### ISSUE TITLE
Create template for bases of unbased super/subscript

### ISSUE THREAD

**Issue by kschaefe** [mathquill/mathquill#467]
The config option `supSubsRequireOperand` prevents the insertion of a superscript or subscript unless there is an operand.  I'd like to be able to have an option that allows an unbased superscript (or subscript) to be inserted with an empty block for the base.

We want to be able to give the user (in our case, math students) the flexibility to insert the power without having type anything else first, but still present valid math (where valid is includes math with unfilled boxes).  Note, I would expect that we would also want this same functionality with subscripts though we don't have a need for that now.  I mostly refer to superscripts below, but you could substitute subscript in most locations.

I had attempted to use `LiveFraction.createLeftOf` as an example for how I might walk backward to consume an already created block and make it part of the new base+power template (git diff below).  This worked very well, giving me the exact behavior that I wanted.  I got 76 test failures.

Most of these arise because the `latex` function will not parse my content correctly.  It shouldn't because the content wants to walk backwards, which the parser shouldn't (and from what I can tell) does not allow.  So, I really need a `LiveSuperscript`, but LaTeX reserves `^` making that option moot.

What would be the best way to accomplish this?  Continuing down the current path of keeping the `latex` parsing as is and using the new work for *live* edits?  This would result in DOM that is different, such that `latex('x^2')` would have a different DOM representation than typing (or using `typedText`).  It should *look* the same, but it won't be the same.

**Comment by kschaefe** [comment]
```diff
diff --git a/src/commands/math/commands.js b/src/commands/math/commands.js
index 5a96f08..4859ba3 100644
--- a/src/commands/math/commands.js
+++ b/src/commands/math/commands.js
@@ -142,8 +142,34 @@ var Class = LatexCmds['class'] = P(MathCommand, function(_, super_) {
 var SupSub = P(MathCommand, function(_, super_) {
   _.ctrlSeq = '_{...}^{...}';
   _.createLeftOf = function(cursor) {
-    if (!cursor[L] && cursor.options.supSubsRequireOperand) return;
-    return super_.createLeftOf.apply(this, arguments);
+    if (!cursor[L]) {
+      if (cursor.options.supSubsRequireOperand) return;
+    }
+
+    if (!this.replacedFragment) {
+      var leftward = cursor[L];
+      while (leftward &&
+        !(
+        leftward instanceof BinaryOperator ||
+        leftward instanceof (LatexCmds.text || noop) ||
+        leftward instanceof SummationNotation ||
+        leftward.ctrlSeq === '\\ ' ||
+        /^[,;:]$/.test(leftward.ctrlSeq)
+        ) //lookbehind for operator
+      ) leftward = leftward[L];
+
+      if (leftward instanceof SummationNotation && leftward[R] instanceof SupSub) {
+        leftward = leftward[R];
+        if (leftward[R] instanceof SupSub && leftward[R].ctrlSeq != leftward.ctrlSeq)
+          leftward = leftward[R];
+      }
+
+      if (leftward !== cursor[L]) {
+        this.replaces(Fragment(leftward[R] || cursor.parent.ends[L], cursor[L]));
+        cursor[L] = leftward;
+      }
+    }
+    super_.createLeftOf.apply(this, arguments);
   };
   _.contactWeld = function(cursor) {
     // Look on either side for a SupSub, if one is found compare my
@@ -290,10 +316,12 @@ LatexCmds.subscript =
 LatexCmds._ = P(SupSub, function(_, super_) {
   _.supsub = 'sub';
   _.htmlTemplate =
-      '<span class="mq-supsub mq-non-leaf">'
-    +   '<span class="mq-sub">&0</span>'
-    +   '<span style="display:inline-block;width:0">&#8203;</span>'
-    + '</span>'
+       '<span class="mq-non-leaf">'
+    +    '<span class="mq-base">&0</span>'
+    +    '<span class="mq-supsub mq-non-leaf">'
+    +      '<span class="mq-sub">&1</span>'
+    +    '</span>'
+    +  '</span>'
   ;
   _.textTemplate = [ '_' ];
   _.finalizeTree = function() {
@@ -308,8 +336,11 @@ LatexCmds.supscript =
 LatexCmds['^'] = P(SupSub, function(_, super_) {
   _.supsub = 'sup';
   _.htmlTemplate =
-      '<span class="mq-supsub mq-non-leaf mq-sup-only">'
-    +   '<span class="mq-sup">&0</span>'
+      '<span class="mq-non-leaf">'
+    +   '<span class="mq-base">&0</span>'
+    +   '<span class="mq-supsub mq-non-leaf mq-sup-only">'
+    +     '<span class="mq-sup">&1</span>'
+    +   '</span>'
     + '</span>'
   ;
   _.textTemplate = [ '**' ];
```

**Comment by laughinghan** [comment]
> the content wants to walk backwards, which the parser [...] does not allow

Yeah, which is a problem independent of this feature, as there's currently no way for a command to implement `\over` or `\choose`. And the parser is monkey-balls slow. It definitely needs to be overhauled. I've been thinking that maybe somehow both problems could be solved if instead of the parser combinator methods that pretty much form a combinatorial algebra, we went back to explicit recursive descent but somehow when delegating parsing to commands, pass in the entire tree so far so that the command can mutate it however it wants (e.g. walk backwards). Maybe it could be a little like FastParse's internally mutable result: http://lihaoyi.github.io/fastparse/#ExternalImmutabiliy

> new base+power template

The empty block will show up fine if you do this, but isn't it weird to be able to type next to the empty block? Like, do you want it to be possible to type `x{ }^2`? Even worse, once you've typed something in the base block and you move leftwards, won't there be like an invisible barrier between being inside and outside the base? Like if you type `a+x^2` and then delete the `+`, when moving left and right there'll be an invisible thing to jump over between `a` and `x`, right?

**Comment by laughinghan** [comment]
Perhaps deciding whether to show a gray box or not in `.sibling{Created,Deleted}()` makes more sense? Then you can have `a+^2` show an "empty block" no matter how it's typed or rendered from LaTeX.

**Comment by kschaefe** [comment]
Yeah, there were definitely more potential problems there than just failing to parse and present the LaTeX.  I am still trying to get a handle on all of the many bits here and was going by analogy to solve the problem replicating LiveFraction.

I appreciate the feedback.  I'll check it `sibling`.  Are there good examples in the code of how to use it?

**Comment by laughinghan** [comment]
Yeah no I didn't mean to imply your approach was bad, just pointing out the
first nontrivial problem to come to mind. This is a pretty interesting
idea, and actually kind of similar to one that I've heard before, at one
point someone from Khan Academy discussed having stuff like plus, minus,
times, and divide (as in \div) show gray boxes if they were missing
operands. The sibling hooks didn't exist yet, though, and since their
initial attempt was a similar approach to yours, with actual blocks, my
response was "oh, that's really more like a semantic tree like Content
MathML, whereas MathQuill's edit tree is a syntax tree like Presentation
MathML because that maps better to LaTeX, and while it could be cool to
adapt MathQuill to edit a completely differently-shaped, semantic rather
than syntactic edit tree, that'd be a huge endeavor and there are more
pressing problems I'd like to solve" and we kinda dropped it.

The "classic" use for the sibling hook is PlusMinus, to dynamically respace
plusses and minuses depending on whether they're being used as unary or
binary operators (when this was the hook's only use, it was actually called
"respace"). An (ab?)use for the sibling hook is Letter, which
auto-unitalicizes "sin", "log" etc.

Han

**Comment by kschaefe** [comment]
Well, the `sibling` suggestion put me on the right path...I think.  I have added an initial check-in of [my changes](https://github.com/kschaefe/mathquill/commit/c9c7b2a43af407226d53bf1b850d950ecc703d3e).  They are rather successful at using the :before pseudo-element to display the template box.

As the commit message states, I am having a couple of problems:
1. I need to know the proper way to get access to options to determine if I should be adding the `mq-empty-base` class.
2. I am at a loss for making the pseudo-element work with the subscript, which is always a block and cause layout issues.

**Comment by kschaefe** [comment]
I started fooling around with the JS Bin listed in math.less for super/subscript ideas.  I think, I may have solved the combination problem with [this effort](http://jsbin.com/yuhamapogu/4/edit?html,css,output).  Note the X's in the before content are just to give a visual indicator.

I may need a mq-sub-only class to handle solo subscripts.  Is there a reason this was not done already?

**Comment by laughinghan** [comment]
In this case, the argument `opts` is the options object, so that's easy. (In general there does need to be a more consistent way to access options.)

If you inserted and removed a gray box element on respace, there'd be no problems like where the pseudo-element inherits its parent's styles, right? $.fn.insDirOf() even takes into account commands whose .jQ has more than one element

**Comment by laughinghan** [comment]
I'm really hoping changing the CSS of supsub is unnecessary because _very_ extensive cross-browser testing went into that vertical align stuff

**Comment by kschaefe** [comment]
But `opts` is not passed to the method in all cases.  Should I be testing for existence?

I don't plan on modifying any CSS that already exists, simply adding some cases for the new `mq-empty-base` class that I added.

**Comment by laughinghan** [comment]
Oh you mention that in the commit message. That's only when `contactWeld` calls `respace`, right? That's a bug, it should pass in `cursor.options`.

I see, I was worried you might be modifying CSS because you mentioned a `mq-sub-only` class. Although to be honest I'm not sure why it's not just `.mq-supsub:before`, why `.mq-sup-only`?

**Comment by kschaefe** [comment]
I'll update the locations that don't properly pass `opts`.

The CSS when there is a subscript makes the super and subscript regions into blocks, but with `.mq-sup-only` it is inline-block.  I need different strategies to tackle those problems.

**Comment by kschaefe** [comment]
In running through to fix the missing `options` parameter to `respace` for https://github.com/mathquill/mathquill/issues/467#issuecomment-143514786, I encountered a lot of places where we're passing both `options` and `cursor`.  The former always a reference from the latter.

That smells a little fishy to me.  Also @laughinghan notes in a [previous comment](https://github.com/mathquill/mathquill/issues/467#issuecomment-143504107):
> (In general there does need to be a more consistent way to access options.)

Should I update #456 with this information or track it elsewhere?

**Comment by laughinghan** [comment]
#456 is external-facing problems only, we should track the internal `options` refactor elsewhere

**Comment by laughinghan** [comment]
Yeah, there are a few places where we're passing in both `options` and `cursor`, because for a while I thought I could avoid having a `cursor.options` at all. I guess those redundant `options` arguments should be removed now.

However, for your changes, I think all you need to do is pass in `cursor.options` to `.respace()`, no? https://github.com/kschaefe/mathquill/blob/c9c7b2a43af407226d53bf1b850d950ecc703d3e/src/commands/math/commands.js#L188

**Comment by kschaefe** [comment]
Yes, but `cursor` was `undefined`.  I had to go all the way back to `renderLatex{Math,Text}()` to pass cursor down the tree.

**Comment by laughinghan** [comment]
Oh...rats...it's `undefined` because https://github.com/mathquill/mathquill/blob/dev/src/commands/math.js#L11-L12 (ea7307eb4fac77c149a11ffdf9a831df85247693)

**Comment by kschaefe** [comment]
I have updated my repo with [changes](https://github.com/kschaefe/mathquill/commit/64f8aa7fe102b4ec1bfa9d7e36246346ce40c6f9) that should fix the `undefined` cases that I was encountering.  Verified via the visual and units tests.  Any spots that you can think of that the tests might not capture?  Happy to build a new one if necessary.

**Comment by laughinghan** [comment]
Wait that comment/commit is silly. Unless I'm missing some edge case, conceptually `contactWeld` is stateless, if there's two SupSubs next to each other then they should merge, and if the cursor is next to or in between them then it goes somewhere, and where it goes shouldn't depend on whether we're typing or writing LaTeX.

**Comment by laughinghan** [comment]
The new commit seems fine by me. The tests seem trustworthy enough, SupSub contact-welding is pretty well-tested.

**Comment by kschaefe** [comment]
I have added the ability to now show [unbased subscripts](https://github.com/kschaefe/mathquill/commit/d48722e272936e174088f993a9a19fdb670d881e).  In that check in I added `.mq-sub-only` class to the DOM mostly to make it easier to get the CSS working.  It was workable without, but the new class makes it easier.  It also makes it much easier to find out when there is both a superscript and a subscript with `:not(.mq-sup-only):not(mq-sub-only)`.

The difficulty that I am experiencing is how to show the empty template box when there is both a superscript and a subscript.  While this case is probably very rare, I feel like I need to solve it before this is really a workable solution.  I have tried a few things and been close, but the visuals get very wonky if superscript and subscript are not the same width.

I am also experiencing problems with the DOM `<span style="display:inline-block;width:0">&#8203;</span>` that is attached to subscripts?  What is the purpose of that DOM?  In Chrome, it doesn't seem to do anything except make my life harder.

Also, the current CSS approach for empty blocks
```css
selectors-for-empty- blocks {
  visibility: hidden;
  content: 'c';
}
```
uses real content, which would be accessible via a screen reader.  My approach is
```css
selectors {
  content: '\2002'; // EN SPACE
  white-space: pre;
   color: transparent; //just in case
}
```
Is there a particular reason that you're using `'c'` over some type of space?

**Comment by kschaefe** [comment]
Just found that my `'c'` question was asked in #426, but my space solution performs the safe function.  Is there any known reason not to use the space?

**Comment by laughinghan** [comment]
> Is there any known reason not to use the space?

I don't think so, I think at the time I did the `'c'` thing it just hadn't occurred to me to try `white-space: pre`.

Is there any particular reason you're using the non-ASCII en space rather than just the ASCII space `' '`? I think the comment would be unnecessary if you did that, as should the `color: transparent` because there's no chance of the browser substituting a question-mark-in-a-box or whatever.

> I am also experiencing problems with the DOM `<span style="display:inline-block;width:0">&#8203;</span>` that is attached to subscripts? What is the purpose of that DOM? In Chrome, it doesn't seem to do anything except make my life harder.

It's for vertical alignment of tall subscripts: 
![image](https://cloud.githubusercontent.com/assets/225809/10252687/085656a8-68ee-11e5-81f1-abf4d61fb420.png)

**Comment by kschaefe** [comment]
> Is there any particular reason you're using the non-ASCII en space rather than just the ASCII space ' '? 
> I think the comment would be unnecessary if you did that, as should the color: transparent because 
> there's no chance of the browser substituting a question-mark-in-a-box or whatever.
En-space provides a width very similar to `'c'`, but ' ' is smaller.  It causes less horizontal shifting as a lot of characters are n-width.  Em-space kept collapsing down and space kept being too small.  Honestly, the `transparent` bit is overkill and no browser should need it, but I was being pedantic.  I guess I could use two spaces, since it is `white-space: pre`.

I play around with the tall subscripts and my unbased code.  It might be that I won't need such a beast until there's a base at which point I could put it back in.

**Comment by laughinghan** [comment]
> I guess I could use two spaces, since it is `white-space: pre`.

I'd prefer that. No comment would be necessary.

> It might be that I won't need such a beast until there's a base at which point I could put it back in.

Even without a base the vertical alignment matters, since there could be text to the right of the subscript, or you might have like `x+{ }_0`.

This looks like great work, by the way. It's crazy how complicated the CSS has to be for such simple features.
